Add regular tree generator

stecrotti opened the pull request:
Generator for a [regular tree](https://en.wikipedia.org/wiki/Bethe_lattice), also known as Bethe Lattice or Cayley Tree, which generalizes a binary tree to arbitrary degree. 

I tried to keep the code as close as possible to the one for `binary_tree` [here](https://github.com/JuliaGraphs/Graphs.jl/blob/ea6bcfeb71de9d1229bbd411ef9b10cac44fa795/src/SimpleGraphs/generators/staticgraphs.jl#L476-L512).

abraunst commented on src/SimpleGraphs/generators/staticgraphs.jl:
````diff
@@ -536,6 +536,68 @@ function double_binary_tree(k::Integer)
+
+# Examples
+```jldoctest
+julia> binary_tree(4)
````
Old doctest?

aurorarossi commented on src/SimpleGraphs/generators/staticgraphs.jl:
```diff
@@ -536,6 +536,54 @@ function double_binary_tree(k::Integer)
+    k <= 0 && return SimpleGraph(0)
+    k == 1 && return SimpleGraph(1)
+    if Graphs.isbounded(k) && BigInt(z)^k - 1 > typemax(k)
+        throw(DomainError(k, "z^k - 1 not representable by type $T"))
```
Here I would change the error from DomainError to InexactError.
I think the error is similar to:
```jl
julia> a::Int8=1923458675849328374548392
ERROR: InexactError: trunc(Int8, 1923458675849328374548392)
```
In our case `a` is `z^k`. What do you think?

simonschoelly commented on src/SimpleGraphs/generators/staticgraphs.jl:
````diff
@@ -536,6 +536,54 @@ function double_binary_tree(k::Integer)
+```
+"""
+function regular_tree(k::T, z::T) where {T<:Integer}
+    k <= 0 && return SimpleGraph(0)
+    k == 1 && return SimpleGraph(1)
````
In these cases we return a graph of of eltype `Int`, but in all other cases we return a graph of type `T`.

simonschoelly replied:
To be completely honest, I am not sure if it is the best idea to determine the eltype of the graphs from `k` and `z` - we do that indeed for other graphs here, but I never really liked that.

simonschoelly replied:
Maybe we can have a function signature
```julia
regular_tree(T::Type{<:Integer}, k::Integer, z::Integer)
```
and then define something like
```julia
regular_tree(k::Integer, z::Integer) = regular_tree(Int64, k, z)
```
that for be similar to the function `sprand` from `SparseArrays` for example.

simonschoelly commented on src/SimpleGraphs/generators/staticgraphs.jl:
```diff
@@ -536,6 +536,54 @@ function double_binary_tree(k::Integer)
+"""
+    regular_tree(k::Integer, z::integer)
+
+Create a [regular tree](https://en.wikipedia.org/wiki/Bethe_lattice), 
```
That wikipedia article might be a bit confusing, because it talks about infinite trees in the introduction.

Maybe we can decribe that a bit more clearly here?

If I understood correctly, what you want to create is a perfect k-ary tree such as described here? https://en.wikipedia.org/wiki/M-ary_tree#Types_of_m-ary_trees

stecrotti commented:
By the way I called it 'regular tree' because of the analogy with regular graphs but that might not be too accurate because:
1. A regular tree is not a regular graph (because of the leaves)
2. z-regular graphs have nodes of degree z while z-ary trees have nodes of degree z+1, potentially confusing

That said, `mary_tree(k, m)`, `zary_tree(k,  z)` don't look great to me.

stecrotti commented:
Changes:
- New method with type `regular_tree(T::Type{<:Integer}, k::Integer, z::Integer)`
- Improved docs with correct wiki page
- `InexactError` instead of `DomainError` when number of vertices is not representable
- Assert z > 0, fallback to a path graph in case z = 1
- Appropriate conversions to the provided type `T`, checked by test

aurorarossi commented on src/SimpleGraphs/generators/staticgraphs.jl:
```diff
@@ -536,6 +536,60 @@ function double_binary_tree(k::Integer)
+    end
+    return SimpleGraph(ne, fadjlist)
+end
+
```
I would add the docstring also to this function. 
````suggestion

"""
    regular_tree(k, z)

Create a regular tree or [perfect z-ary tree](https://en.wikipedia.org/wiki/M-ary_tree#Types_of_m-ary_trees): 
a `k`-level tree where all nodes except the leaves have exactly `z` children. 
For `z = 2` one recovers a binary tree.
# Examples
```jldoctest
julia> regular_tree(4, 3)
{40, 39} undirected simple Int64 graph
julia> regular_tree(5, 2) == binary_tree(5)
true
```
"""
````

simonschoelly replied:
It's quite often the case, that the docstrings for multiple methods of a function are only added for a single method - I think that would also be better here, than adding docstrings for both methods of this function.

aurorarossi reviewed (commented):
Some comments on the docstrings. If you agree with the style then I will open a PR to fix `double_binary_tree` and `binary_tree`

dapias commented:
I just realized that a $k$-ary tree is not the same as a Cayley tree of connectivity $k$ because in a $k$-ary tree, the root has connectivity $k$, and all the nodes excepting those in the leaves have connectivity $k+1$. Look, for instance, at the attached plot, nodes 2,3 and 4 have connectivity four and not three as it would be the case for a Cayley tree

<img width="474" alt="Bildschirm­foto 2022-12-14 um 09 31 30" src="https://user-images.githubusercontent.com/5333281/207545547-c55ae392-af61-4fba-9f7a-139b23f144e4.png">

stecrotti commented:
Agree! Indeed it was corrected and should be ok in the current version of this PR

dapias commented:
Is there a function to create an actual Cayley tree?

simonschoelly commented on src/SimpleGraphs/generators/staticgraphs.jl:
````diff
@@ -536,6 +536,60 @@ function double_binary_tree(k::Integer)
+```
+"""
+function regular_tree(T::Type{<:Integer}, k::Integer, z::Integer)
+    z <= 0 && throw(DomainError(z, "number of children must be positive"))
````
Shouldn't we also check for `k >= 0` here?

simonschoelly commented on src/SimpleGraphs/generators/staticgraphs.jl:
```diff
@@ -536,6 +536,60 @@ function double_binary_tree(k::Integer)
+    z == 1 && return path_graph(T(k))
+    k <= 0 && return SimpleGraph(zero(T))
+    k == 1 && return SimpleGraph(one(T))
+    if Graphs.isbounded(k) && (BigInt(z)^k - 1) ÷ (z - 1) > typemax(T)
+        throw(InexactError(:convert, T, (BigInt(z)^k - 1) ÷ (z - 1)))
+    end
+
+    n = T((z^k - 1) / (z - 1))
```
You already calculate the number of vertices with `(BigInt(z)^k - 1) ÷ (z - 1)`, so we we could just reuse this for `n`.

With the expression `n = T((z^k - 1) / (z - 1))` there are two issues:
- `z^k` might overflow before converting it to `T`
- By using the `/` operator instead of `÷` your intermediate result will be a double

simonschoelly replied:
A lot of these overflow issues probably won't happen if we restrict  `k` and `z` to `Int` in the function definition.

stecrotti commented:
Changed the signature to match the rest of the file as suggested by @aurorarossi, but kept only one docstring.

Fixed potential overflow as suggested by @simonschoelly.

simonschoelly commented:
There is this one comment from me, about `k >= 0`  - not sure if you have seen it, but feel free to ignore that it you think we should not change anything there.

Otherwise looks good for me - I did not check the math, but choose to believe the tests.

stecrotti commented:
Sure, i replied under your review comment: 
> I stuck to what was used in the rest of the file which is that for k<=0 it returns a SimpleGraph(zero(T)). I guess there was a good reason for that? If not it makes total sense to error if k<0.
